File: solutions/_61_rotate_list/solution.py
from typing import Optional


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def rotateRight(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        if not head:
            return head

        length = 1
        curr = head
        while curr.next:
            curr = curr.next
            length += 1

        k = k % length
        if k == 0:
            return head

        curr.next = head
        for _ in range(length - k):
            curr = curr.next

        new_head = curr.next
        curr.next = None

        return new_head


# https://leetcode.com/problems/rotate-list/

File: solutions/_61_rotate_list/Solution.java

package _61_rotate_list;

class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        if (head == null) {
            return null;
        }
        int len = 1;
        ListNode curr = head;
        while (curr.next != null) {
            curr = curr.next;
            len++;
        }
        k = k % len;
        if (k == 0) {
            return head;
        }

        curr.next = head;
        for (int i = 0; i < len - k; i++) {
            curr = curr.next;
        }
        head = curr.next;
        curr.next = null;
        return head;
    }
}

/*
 * https://leetcode.com/problems/rotate-list/
 */